### PR TITLE
Relax the label key validation

### PR DESCRIPTION
**Pull request by gunjan5**
- allows 0 to 253 chars
- alphanumeric, `-`, `_`, `/`, and `.`
- end with only alphanumeric char

### CODE REVIEW

**Comment by gunjan5** [comment]
Fixes #322

**Comment by gunjan5** [comment]
CC @robbrockbank @heschlie

File: lib/validator/validator.go [hunk 1]
```diff
@@ -34,7 +34,7 @@ var validate *validator.Validate
 var (
 	nameRegex          = regexp.MustCompile("^[a-zA-Z0-9_.-]{1,128}$")
 	interfaceRegex     = regexp.MustCompile("^[a-zA-Z0-9_-]{1,15}$")
-	labelRegex         = regexp.MustCompile("^([a-zA-Z0-9.-]{0,253}/)?[a-zA-Z0-9]([a-zA-Z0-9_.-]{0,61}[a-zA-Z0-9])?$")
+	labelRegex         = regexp.MustCompile(`^([a-zA-Z0-9\_\.\-\/]{0,252}[a-zA-Z0-9])?$`)
```
Comment (caseydavenport) [inline lib/validator/validator.go]: Why did you change from quotes to backticks?

Also, do we really need to escape all of those?

Comment (caseydavenport) [inline lib/validator/validator.go]: Also, why the switch to 252 characters?

**Comment by heschlie** [comment]
@gunjan5 I believe you need to make a similar change here as well https://github.com/projectcalico/libcalico-go/blob/master/lib/selector/tokenizer/tokenizer.go#L57

File: lib/validator/validator.go (hunk 1 above)
Comment (gunjan5) [inline lib/validator/validator.go]: > Why did you change from quotes to backticks?

because you get a compile time error with some of those escape chars, so have to use runes.

> Also, do we really need to escape all of those?

yeah the only one that could have been without escape was `/` but I did it to be consistent.

> Also, why the switch to 252 characters?

so it's 252 (alphanumeric+specials) + 1 only alphanumeric so it doesn't end with a special char

Comment (caseydavenport) [inline lib/validator/validator.go]: > yeah the only one that could have been without escape was / but I did it to be consistent.

Interesting, what is the behavior of `_`, for example, when not escaped?

Also, no need to be consistent - it's more confusing the unnecessarily escape something (might come back later and be confused by it)

Comment (caseydavenport) [inline lib/validator/validator.go]: As should `-`, I think.

Also, looks like `.` doesn't need to be escaped either: https://regex101.com/r/PH8aQM/1

Comment (caseydavenport) [inline lib/validator/validator.go]: > so it's 252 (alphanumeric+specials) + 1 only alphanumeric so it doesn't end with a special char

The old format was 253 alphanumeric+specials followed by a slash + 61  alphanumric+specials followed by an alphanumeric.  Right?

So, we've gone from total length 253+1+61+1 to 252+1?

Comment (caseydavenport) [inline lib/validator/validator.go]: What about this regex?
```
^([a-zA-Z0-9_.-/]{0,253}/)?([a-zA-Z0-9][a-zA-Z0-9_.-]{0,61}[a-zA-Z0-9])?$
```

This is essentially the same as before, but it supports slashes and underscores in the prefix, and has two splits:
- The prefix.
- The value after the prefix.

For comparison:

```
^([a-zA-Z0-9_.-/]{0,253}/)?([a-zA-Z0-9][a-zA-Z0-9_.-]{0,61}[a-zA-Z0-9])?$
^([a-zA-Z0-9.-]{0,253}/)?[a-zA-Z0-9]([a-zA-Z0-9_.-]{0,61}[a-zA-Z0-9])?
```

**Comment by gunjan5** [comment]
Fixes #307 as well

File: lib/validator/validator.go [hunk 2]
```diff
@@ -26,16 +26,18 @@ import (
...
 	"github.com/projectcalico/libcalico-go/lib/selector"
+	"github.com/projectcalico/libcalico-go/lib/selector/tokenizer"
+
 	"gopkg.in/go-playground/validator.v8"
 )
 
 var validate *validator.Validate
 
 var (
+	labelRegex         = regexp.MustCompile(`^` + tokenizer.LabelKeyMatcher + `?$`)
```
Comment (caseydavenport) [inline lib/validator/validator.go]: Hm. I think the `?` belongs in the LabelKeyMatcher.  Right?

The `?` means that the trailing `([a-zA-Z0-9_.-]{0,61}[a-zA-Z0-9])` can occur once or not at all.

File: lib/validator/validator_test.go [hunk 3]
```diff
@@ -113,14 +113,12 @@ func init() {
...
 		Entry("should accept label key with dns prefix", api.HostEndpointMetadata{Labels: map[string]string{"calico/k8s_ns": "kube-system"}}, true),
-		Entry("should accept label key where prefix is 253 characters", api.HostEndpointMetadata{Labels: map[string]string{"Projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org12/k8s_ns": "gold"}}, true),
-		Entry("should accept label key where name segment is 63 character", api.HostEndpointMetadata{Labels: map[string]string{"projectcalico.org/gold._0-9gold._0-9gold._0-9gold._0-9gold._0-9gold._0-9gold._012": "gold"}}, true),
+		Entry("should accept label key where prefix is 253 characters", api.HostEndpointMetadata{Labels: map[string]string{"projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.projectcalico.org.calico12345/k8s_ns": "gold"}}, true),
```
Comment (caseydavenport) [inline lib/validator/validator_test.go]: Looks like we no longer test a key that starts with a capital?

Comment (gunjan5) [inline lib/validator/validator_test.go]: not sure if that makes any difference. Most keys are not capitalized anyway

File: lib/validator/validator.go (hunk 2 above)
Comment (gunjan5) [inline lib/validator/validator.go]: you're absolutely right, changed it

File: lib/validator/validator_test.go (hunk 3 above)
Comment (caseydavenport) [inline lib/validator/validator_test.go]: I think it does.  We don't want to accidentally break the regex such that it doesn't accept those types.

Should be easy enough to add another entry.

Comment (gunjan5) [inline lib/validator/validator_test.go]: ok done

**Comment by gunjan5** [comment]
ok I finally got it working with felix master. Calico node was using an old felix hash somehow. Tested it with stars and advanced policy demos. @caseydavenport should we merge this and pin it for a felix RC? or cut a libcalico-go v1.0.1 and then felix?

**Comment by gunjan5** [comment]
/CC @fasaxc or @neiljerram can do a Felix release once I get a 👍 from Casey, and this is merged

**Comment by caseydavenport** [comment]
LGTM
